### PR TITLE
nixos/tailscale: add extraUpFlags option

### PR DESCRIPTION
**Pull request by jwygoda**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by colemickens** [comment]
I feel like `extraUpFlags` is a better name than `autoconnectOptions`?

**Comment by MinerSebas** [comment]
The PR Title also needs to be updated from `autoconnectOptions` to `extraUpFlags`.

**Review by drupol — COMMENTED** [review]
I'm okayish with this PR... but given the small amount of extra flags we can pass to `tailscale`, I'm wondering if it woudn't be a better idea to just add something like:

```nix
runSSh = mkOption {
  description = lib.mdDoc "run an SSH server, permitting access per tailnet admin's declared policy";
  type = types.bool;
  default = false;
};
```

WDYT ?

File: nixos/modules/services/networking/tailscale.nix [hunk 1]
```diff
@@ -58,6 +58,13 @@ in {
         A file containing the auth key.
       '';
     };
+
+    extraUpFlags = mkOption {
+      description = lib.mdDoc "Extra flags to pass to {command}`tailscale up`.";
+      type = types.separatedString " ";
```
Comment (Kranzes) [inline nixos/modules/services/networking/tailscale.nix:64]: please make this a `listOf str` and then use `escapeShellArgs` to turn it into one unified string.

**Comment by jwygoda** [comment]
Thanks for the reviews.
@drupol There are over [20 flags](https://tailscale.com/kb/1241/tailscale-up/) we can pass to `tailscale up` so I'm going to go with `extraUpFlags`.
@Kranzes Done, also changed `pkgs.tailscale` to `cfg.package`.

**Review by drupol — APPROVED** [review]
I'm ok with the changes.

However, I would have replaced the `package` attribute with this single one liner:

```nix
  package = lib.mkPackageOptionMD pkgs "tailscale" {
```

Do you think you can add this as well ?

**Review by drupol — APPROVED** [review]
:100: ! Thanks ! Merging this as soon as the CI is done.

**Pull request merged**